### PR TITLE
Fix TargetPosition targeting and give D2k buildings better hit-shapes

### PR DESCRIPTION
**Pull request by reaperrr**
The discussion in #13165 leans towards footprint-based hit-shapes, so I've built on that plus the initial review feedback.
Rather than updating the old PR, I've decided to file this separately (mostly due to the branch name, but also to keep the discussion and actual reviewing a bit separate).

This is basically the spiritual successor to #10383 and actually scavenged some code from there.

The commit count makes this look more intimidating than it really is, only the first 3 commits contain code and the individual yaml commits are small.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/TargetableFootprint.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Actor can be targeted and is a building with a footprint.")]
+	public class TargetableFootprintInfo : TargetableInfo
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/TargetableFootprint.cs]: You use `Trait<Building>`, so this dependency must be declared with a `Requires<BuildingInfo>` here.

**Comment by pchote** [comment]
> For invisible insta-hit projectiles to shoot at the closest target
position instead, modders should either use Bullet or set an Inaccuracy of
>= 1.

Since this requires changing the projectile args IMO it would be clearer and more correct to introduce a specific flag for this than to bolt it on to the inaccuracy.

**Comment by pchote** [comment]
I played with this briefly in TS and TD, and I'm really liking it.  Having the impacts land on the "outside" of the buildings make them feel more solid, and overall improves the polish in a small but hopefully significant way.

One issue that I did notice is that units still face towards the center position, which looks quite wrong when you have a gun turret or rocket trooper standing next to the corner of a construction yard.

**Comment by pchote** [comment]
Moving the explosions away from the center of the structure causes a tricky problem in the classic mods: explosions from behind now render behind the structure instead of on top.  Forcing a +1 cell z offset on all explosions would work around this, but feels a bit hacky.

**Comment by pchote** [comment]
The awkward diagonal perspective on the RA conyard means that impacts land visually outside the structure on the bottom corner cells.  A potential solution is to have another new trait that lets you list the targetable positions directly, and then fudge those corner points further inwards.

**Comment by atlimit8** [comment]
It might be better to create a hit-shape that is a collection of hit-shapes then copy one for each occupied cell. The footprint is used for marking occupied positions rather than the shape. Then use new trait as @pchote mentions, or possibly extend `ITargetable` or `Targetable` to include `IEnumberable<WPos> Positions` with the default being the center (`TargetableInfo` could contain `WVec[] Offsets`).

**Comment by reaperrr** [comment]
Having played around with this a bit locally, I think we should either
a) keep `ITargetablePositions` in a trait separate from `Targetable`, or
b) do as @atlimit8 suggests.

Reasoning: It is both rather annoying and risky wrt regressions on the yaml side to have to disable `TargetableX` and enable `TargetableY` in the rules when you just want to change an inherited target position type and have not intention to change TargetTypes, for example.

**Comment by pchote** [comment]
I think it would make the most sense to merge`ITargetablePositions` into `IHitShape`, and then as suggested have a new type of hit shape for multi-celled actors.

**Comment by reaperrr** [comment]
That approach has two problems (at least within the scope of this PR):
1) We might want some actors to be "targetable", but not have the health trait
2) polygon hitshape and related warhead changes are out of my league

**Comment by atlimit8** [comment]
> I think it would make the most sense to merge `ITargetablePositions` into `IHitShape`

I actually had the same idea afterwards and I did say "Then use *new trait* as @pchote mentions, *or possibly* merge `ITargetablePositions` into `ITargetable`...". Yea, with my idea of using `Exit`s for target-able positions (the genesis of #13208 really), I had abandoned merging into `Targetable` and now would rather targetable position providers specify valid types via conditions and have `Targetable` just act as a manager where at least one position must be valid for a targetable type to be valid:
```yaml
    Targetable:
    Exit@1:
        ...
        ProductionTypes: 
            Vehicle.Repair: !powered-down # can't open door without power
        TargetTypes:
            Vehicle.Repair: !powered-down # can't open door without power
    Health:
        ...
        TargetTypes:
            Vehicle: !powered-down # can't open door without power
```

**Comment by atlimit8** [comment]
Such a refactor could easily handle multiple `Health` traits if they just disable their target types with themselves.

**Comment by atlimit8** [comment]
A convex polygon can be handled by decomposing it into right triangles and rectangles. Perhaps it would be better if I worked on the hit shapes then.

**Comment by reaperrr** [comment]
By the way, one of the reasons why I implemented and think we should continue to offer a way to derive targetable positions from building footprints is that larger buildings simply have a lot of cells, so manually setting targetable positions via WVec lists is awfully inefficient in terms of time. Modders need some automatic, simple way to do that, in my opinion, and occupied cells are the easiest approach in my book.

**Comment by atlimit8** [comment]
I already thought about footprints and have my own idea:
```yaml
SomeBuilding:
    Health:
        Shape: ByFootprint
            Footprint: _x_ xxx axb # If missing, pull from Building
            Shapes: # each shape template operates from the center of the instancing cell
                x: Rectangle
                    TopLeft: -512, -512
                    BottomRight: 512, 512
                a: RightTriangle
                    RightCorner: 512, -512
                    Legs: -1024, 1024
                    TargetableOffset: 256, -256
                    DamageMultiplier: 50
                b: RightTriangle
                    RightCorner: -512, -512
                    Legs: -1024, 1024
                    TargetableOffset: -256, -256
                    DamageMultiplier: 50
    Building:
        ...
        Footprint: _x_ xxx =x=
        ...
```

**Comment by pchote** [comment]
We need to be really careful about having a whole bunch of shapes per object, because we have to enumerate them every time we calculate damage etc.  If we have a normal square footprint then we really shouldn't be using more than a single square shape.

**Comment by atlimit8** [comment]
@pchote, that can be fudged:
```yaml
SomeBuilding:
    Health:
        Shape: ByFootprint
            Footprint: _x_ _c_ axb # If missing, pull from Building
            Shapes: # each shape template operates from the center of the instancing cell
                x: Rectangle
                    TopLeft: -512, -512
                    BottomRight: 512, 512
                c: Rectangle
                    DamagePerCell: true
                    TargetablePerCell: true
                    TopLeft: -1536, -512
                    BottomRight: 1536, 512
                a: RightTriangle
                    RightCorner: 512, -512
                    Legs: -1024, 1024
                    TargetableOffsets: 256, -256
                    DamageMultiplier: 50
                b: RightTriangle
                    RightCorner: -512, -512
                    Legs: -1024, 1024
                    TargetableOffsets: -256, -256
                    DamageMultiplier: 50
    Building:
        ...
        Footprint: _x_ xxx =x=
        ...
OtherBuilding:
    Health:
        Shape: ByFootprint
            Footprint: ___ _a_ _b_ # If missing, pull from Building
                a: Rectangle
                    DamagePerCell: true
                    TargetablePerCell: true
                    TopLeft: -1536, -1536
                    BottomRight: 1536, 512
                b: Rectangle
                    DamagePerCell: true
                    TargetablePerCell: true
                    TopLeft: -1536, -512
                    BottomRight: 1536, 0
                    DamageMultiplier: 50
    Building:
        ...
        Footprint: xxx xxx ===
        ...

**Comment by reaperrr** [comment]
Updated.

Adressed the comment on InstantHit, added a debug overlay and for now (temporarily or not) an additional trait for custom positions and a test-case (D2k Repair Pad).

> We need to be really careful about having a whole bunch of shapes per object, because we have to enumerate them every time we calculate damage etc. If we have a normal square footprint then we really shouldn't be using more than a single square shape.

Yeah, I guess that's ultimately the better approach for most buildings.

Now that I think about it, with custom `TargetablePositions` we could technically set up shapes that are smaller than the footprint and just match the visuals of the building.
In other words, as long as we set the TargetablePositions accordingly, we possibly don't even need to use footprint-based/polygon shapes for the shipping mods or TS (which doesn't even have any asymmetric footprints).

**Comment by atlimit8** [comment]
I think a mix of multi-cell support for hit shapes plus `ByFootprint` provides the best compromise between performance and modder accessibility.

**Comment by reaperrr** [comment]
Updated as discussed on IRC for now. 
The last commit shows that we need #12609 sooner rather than later, although we could temporarily work around that by increasing the default values until we remove the Target(Extra)SearchRadius fields completely.

**Comment by MustaphaTR** [comment]
Looks like units look at center of the target regardless of where they are attacking.
turret targeting

Light factory is targeted from its exit site in original. But looks like it is missing a target position here.
lfac

Top parts of IX Res. Center, Hi-Tech Fac and Heavy Fac. are also targeted in original. I think it is missing in this PR as it is not possible to have those wierdly shaped hit shape is not possible yet.

These doesn't look matching.
hitshapes

Ornithopter is still powerful too powerful.

**Comment by pchote** [comment]
> These doesn't look matching.

An easy fix for these and the RA conyard would be to replace the `RotateToIsometry` flag with a general WAngle rotation angle, and then rotate the shapes to more closely match the structures.

**Comment by reaperrr** [comment]
> Looks like units look at center of the target regardless of where they are attacking instead of the closest target position.

Ah, forgot to fix that.

> Light factory is targeted from its exit site in original. But looks like it is missing a target position here.

The original behavior looks bad, though. I'll add a target position on the edge of the exit (at the border of the visible sprite), that should be a good compromise.

> Top parts of IX Res. Center, Hi-Tech Fac and Heavy Fac. are also targeted in original. I think it is missing in this PR as it is not possible to have those wierdly shaped hit shape is not possible yet.

Yes, that was intentional. In my opinion this is good enough until someone like @atlimit8 implements RectangleUnion or Polygon hitshapes, then we can adapt those edge cases. I wanted to avoid using my hacky per-cell hit-shapes, as that hack will eventually go away anyway.

> These doesn't look matching.

I know, I took visual appearance of the buildings into account a bit.

> An easy fix for these and the RA conyard would be to replace the `RotateToIsometry` flag with a general WAngle rotation angle, and then rotate the shapes to more closely match the structures.

I don't think that will help on the Ix Research Center, but for the heavy factory and maybe high tech factory it might work, yeah.

**Comment by reaperrr** [comment]
Depends on #13236 (to revert the custom radius changes) and #13237 (to rotate shapes).

**Comment by reaperrr** [comment]
Rebased and updated.

**Comment by pchote** [comment]
Needs a rebase again.

**Comment by reaperrr** [comment]
Rebased and updated.

I fine-tuned some of the shapes and targetable locations a bit more, it now looks like this:

shapes1
shapes2

**Comment by pchote** [comment]
I think it's a bit odd to have the heavy factory footprint rotated, but none of the other rotated rectangular structures.

**Comment by reaperrr** [comment]
> I think it's a bit odd to have the heavy factory footprint rotated, but none of the other rotated rectangular structures.

Yeah, I was unsure about it, too.
The problem with rotating the shape in the pre-TS legacy mods is that the camera angle Westood used for rendering makes it impossible to use rotated rectangles without either a) having the corners of the hit-shapes exceed both footprint and sprite or b) be too small compared to non-rotated, footprint-based shapes of the other buildings.

Personally, I'd vote for dropping the rotation and using the same shape as the hightech factory and Ix RC, and only tweak the targetable locations to be placed both inside the shape and on the sprite. In my opinion that would be good enough for now.

To fix the shapes properly, we'll ultimately need either the `RectangleUnionShape` @atlimit8 suggested or the completely custom `PolygonShape` we've talked about in the past, but that's both out of my league and would make this PR too large, in my opinion.

**Comment by pchote** [comment]
My vote is to make the heavy factory and also refinery use square footprints, and then we can add a skew parameter to account for the camera angle to both rectangles and circles in a followup PR.

**Comment by reaperrr** [comment]
Updated.
shapes3

File: OpenRA.Mods.Common/HitShapes/Circle.cs [hunk 2]
```diff
@@ -48,25 +51,27 @@ public WDist DistanceFromEdge(WVec v)
 
 		public WDist DistanceFromEdge(WPos pos, Actor actor)
 		{
-			var actorPos = actor.CenterPosition;
+			var orientation = actor.Orientation;
+			var actorPos = actor.CenterPosition + LocalOffset.Rotate(orientation);
 
 			if (pos.Z > actorPos.Z + VerticalTopOffset)
-				return DistanceFromEdge(pos - (actorPos + new WVec(0, 0, VerticalTopOffset)));
+				return DistanceFromEdge((pos - (actorPos + new WVec(0, 0, VerticalTopOffset))).Rotate(-orientation));
```
Comment (pchote) [inline OpenRA.Mods.Common/HitShapes/Circle.cs]: Correcting for the `LocalOffset` here but not in `DistanceFromEdge(WVec v)` smells like a bug.  I would expect to see method calculate the offset relative to center of the actor and then rotate it into the local coordinates, and then `DistanceFromEdge(WVec v)` to account for the `LocalOffset`.

Comment (pchote) [inline OpenRA.Mods.Common/HitShapes/Circle.cs]: The three points of interest (center of actor, center of actor + `v`, center of actor + `LocalOffset`) define a triangle, so you can use the [cosine rule](https://en.wikipedia.org/wiki/Law_of_cosines) to simplify the distance calculation in the other `DistanceFromEdge` overload.

Comment (pchote) [inline OpenRA.Mods.Common/HitShapes/Circle.cs]: I don't think this is necessary for the current state of this PR, so if that sounds too scary then it might be more efficient to drop this commit for now.

File: OpenRA.Mods.Common/Activities/Attack.cs [hunk 3]
```diff
@@ -113,7 +114,8 @@ protected virtual Activity InnerTick(Actor self, AttackBase attack)
 				return NextActivity;
 			}
 
-			var desiredFacing = (Target.CenterPosition - self.CenterPosition).Yaw.Facing;
+			var targetedPosition = Target.Positions.PositionClosestTo(pos);
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Attack.cs]: Turreted units work differently again, so you also need to update `Turreted.FaceTarget`.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 4]
```diff
@@ -679,6 +679,20 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
...
 
+				// Fixed position targeting and added TargetableLocations, while removing ITargetablePositions from Building
+				if (engineVersion < 20170514)
+				{
+					var building = node.Value.Nodes.FirstOrDefault(n => n.Key == "Building");
+					if (building != null)
+					{
+						Console.WriteLine("The Building trait no longer provides per-cell targetable positions,");
+						Console.WriteLine("and weapons no longer unconditionally target actor CenterPosition.");
+						Console.WriteLine("You may need to manually set custom target locations or update building hit shapes.");
```
Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Just pointing out for others that the changes in this PR mean that buildings in ra/cnc/ts will now be targeted only at their CenterPosition.  This will reintroduce several old bugs relating to attack ranges.  By merging this PR we are committing to running the hitshape changes through the other mods, which will be great for the long term but painful for balancing in the short term.

File: OpenRA.Game/Traits/Target.cs [hunk 5]
```diff
@@ -145,12 +145,11 @@ public IEnumerable<WPos> Positions
...
 
-						var targetablePositions = actor.TraitOrDefault<ITargetablePositions>();
-						if (targetablePositions != null)
+						var targetablePositions = actor.TraitsImplementing<ITargetablePositions>().Where(Exts.IsTraitEnabled);
+						if (targetablePositions.Any())
 						{
-							var positions = targetablePositions.TargetablePositions(actor);
-							if (positions.Any())
-								return positions;
+							var target = this;
```
Comment (reaperrr) [inline OpenRA.Game/Traits/Target.cs]: From #10383:
https://github.com/OpenRA/OpenRA/pull/10383#r64640433

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs (hunk 4 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: That was actually one of the reasons why I suggested the bugfix release, to buy us a bit more time for that.

File: OpenRA.Mods.Common/HitShapes/Circle.cs (hunk 2 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/HitShapes/Circle.cs]: Yeah, sorry, with my level of math that's better left for a follow-up then.

**Comment by MustaphaTR** [comment]
I'm not sure if it is a good idea to give HFac 3x3 hitshape while Hi-Tech Fac and IX RC keep 3x2.

Also, saying again, Ornithopters. As now all 3 hit the building well, they deal too much damage to big structures. I'm not sure how exactly to balance that, but i think it would be better to do here unless someone wants to take proper Ornithopter system before Next release.

**Comment by reaperrr** [comment]
> I'm not sure if it is a good idea to give HFac 3x3 hitshape while Hi-Tech Fac and IX RC keep 3x2.

@pchote suggested a follow-up that allows to 'skew' rectangular shapes to better fit the rendered sprite. So this is only temporary either way.

I agree that Ornithopters need to be rebalanced, but I'd prefer to do that in a separate PR after this has been merged.

File: OpenRA.Mods.Common/Activities/Attack.cs (hunk 3 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/Activities/Attack.cs]: Fixed.

**Comment by pchote** [comment]
> I'm not sure if it is a good idea to give HFac 3x3 hitshape while Hi-Tech Fac and IX RC keep 3x2.

Somewhere earlier in the discussion it was pointed out that the original d2k uses full cells for its targeting and hitshapes, like this PR is doing.  It should be easy to check if it indeed used 3x3 for HFac and 3x2 for the others.

**Comment by reaperrr** [comment]
@pchote: All 3 'hit-shapes' were occupancy-based in the original, so `_x_ xxx xxx` is what they'd need in theory.

**Comment by reaperrr** [comment]
Split the easier part of the PR to #13314 (with some updates), to make this here a bit easier to review later.

**Comment by reaperrr** [comment]
Rebased and updated.

All dependencies are merged and this contains the last two code commits, after that setting up hit-shapes and targetable offsets for the other 3 mods is all that remains.

File: mods/d2k/rules/structures.yaml [hunk 6]
```diff
@@ -772,6 +827,13 @@ repair_pad:
 		Dimensions: 3,3
 	Health:
 		HP: 3000
+	BodyOrientation:
+		CameraPitch: 192
+	HitShape:
+		UseOccupiedCellsOffsets: false
+		TargetableOffsets: 0,0,0, 960,0,0, 0,960,0, -960,0,0, 0,-960,0, 704,704,0, -704,-704,0, -704,704,0, 704,-704,0
+		Type: Circle
```
Comment (pchote) [inline mods/d2k/rules/structures.yaml]: This probably wants to use the footprint too now for consistency

**Review by pchote — APPROVED** [review]
Just one nit/question, otherwise looks great :+1:

**Comment by reaperrr** [comment]
Updated.

File: mods/d2k/rules/defaults.yaml [hunk 7]
```diff
@@ -336,8 +336,13 @@
 		VisibilityType: CenterPosition
 	Targetable:
 		TargetTypes: Ground, C4, Structure
+	BodyOrientation:
+		CameraPitch: 255
```
Comment (pchote) [inline mods/d2k/rules/defaults.yaml]: Like RA, this breaks the weapon turning circles.

Comment (reaperrr) [inline mods/d2k/rules/defaults.yaml]: Ah, forgot about that. Fixed.

Comment (pchote) [inline mods/d2k/rules/defaults.yaml]: Why are you overriding this in the first place?

Comment (reaperrr) [inline mods/d2k/rules/defaults.yaml]: The camera pitch affects the custom `TargetableOffsets`, which would be far more difficult to set with default pitch (512 would be far less than half a cell along the vertical axis).

Comment (pchote) [inline mods/d2k/rules/defaults.yaml]: The tradeoff is that this then breaks everything else that relies on having the correct-for-artwork/correct-for-mod camera pitch.  Finding and then scaling the y offsets of the targetable positions is far preferrable to that.

Comment (pchote) [inline mods/d2k/rules/defaults.yaml]: The scaling factor you need to apply for d2k is 64 / 39.

**Comment by reaperrr** [comment]
Updated.

**Review by obrakmann — APPROVED** [review]
lgtm, 👍

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ee74ba4b8f96fc4eee822c3939cbf873d13624d3)
